common/src/chain/config/emission_schedule.rs

```diff
@@ -0,0 +1,611 @@
+        (self.supply_at(ht) - self.supply_at(prev_ht)).expect("Supply not monotonic")
+    }
+}
+
+impl std::fmt::Debug for EmissionSchedule {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "EmissionSchedule(<function>)")
+    }
+}
+
+/// Emission schedule where supply is a piecewise linear function, represented as a table.
+///
+/// The table has a string representation, described in [Self::from_str].
+#[derive(Eq, PartialEq, Debug, Clone)]
+pub struct EmissionScheduleTabular {
+    /// The initial supply, in MLTs
+    initial_supply: Mlt,
+    /// The initial per-block reward, before the first period kicks in.
+    initial_reward: Mlt,
+    /// Reward periods. Specified by block height and the per-block reward starting at that height.
+    periods: BTreeMap<BlockHeight, Mlt>,
+}
+
+impl EmissionScheduleTabular {
+    /// Create a new piecewise-linear supply schedule
+    ///
+    /// This constructor takes:
+    ///  * The initial supply (premine amount)
+    ///  * The initial per-block reward, before the first explicitly specified period (below).
+    ///  * Reward periods. These are mappings from block heights to per-block rewards from that
+    ///    point onwards, until the next period. The periods are implicitly sorted by BTreeMap.
+    pub fn new(
+        initial_supply: Mlt,
+        initial_reward: Mlt,
+        periods: BTreeMap<BlockHeight, Mlt>,
+    ) -> Self {
+        Self {
+            initial_supply,
+            initial_reward,
+            periods,
+        }
+    }
+
+    /// All reward periods, starting at block height 0
+    pub fn reward_periods(&self) -> impl Iterator<Item = (BlockHeight, Mlt)> + '_ {
+        std::iter::once((BlockHeight::zero(), self.initial_reward))
+            .chain(self.periods.iter().map(|(ht, mlt)| (*ht, *mlt)))
+    }
+
+    /// Get the initial supply
+    pub fn initial_supply(&self) -> Mlt {
+        self.initial_supply
+    }
+
+    /// Get tail emission block height and per-block amount
+    pub fn tail_emission(&self) -> (BlockHeight, Mlt) {
+        let empty_tail = || (BlockHeight::zero(), self.initial_reward);
+        self.periods.iter().next_back().map_or_else(empty_tail, |(ht, mlt)| (*ht, *mlt))
+    }
+
+    /// Final supply. `None` if the supply increases indefinitely
+    pub fn final_supply(&self) -> Option<Mlt> {
+        let (tail_block, tail_reward) = self.tail_emission();
+        (tail_reward == Mlt::ZERO).then(|| self.schedule().supply_at(tail_block))
+    }
+
+    /// Get the final emission schedule
+    pub fn schedule(&self) -> EmissionSchedule {
+        let table_seed = (
+            self.initial_supply,
+            self.initial_reward,
+            BlockHeight::zero(),
+        );
+
+        // Pre-calculate a table with starting supply and per-block rewards for each reward period.
+        let table: BTreeMap<BlockHeight, (Mlt, Mlt)> = self
+            .periods
+            .iter()
+            .scan(
```

Comment:
Well, I'm allergic to scan because I'm allergic to mutable... I would've used rfold for this, but sure. If this works...

Reply:
Well conceptually scan is the right operation to use here. I took a brief look into `itertools` to see whether there is a version of `scan` that lets you return the updated values rather than mutating them (which I would prefer too) but unfortunately did not find anything. Anyway all mutation is contained in the closure body in the next 5 lines of code so hopefully this is not so bad.